Changed config to user 'folder'/'files' collections types

gsidebo opened the pull request:
#### Pre-Flight checklist

- [ ] Screenshots and design review for any changes that affect layout or styling
  - [ ] Desktop screenshots
  - [ ] Mobile width screenshots
  - [ ] Tag @ferdi or @pdpinch for review
- [ ] Migrations
  - [ ] Migration is backwards-compatible with current production code
- [ ] Testing
  - [ ] Code is tested
  - [ ] Changes have been manually tested
- [ ] Settings
  - [ ] New settings are documented and present in `app.json`
  - [ ] New settings have reasonable development defaults, if applicable
  - [ ] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?
(Required)

#### What's this PR do?
(Required)

#### How should this be manually tested?
(Required)

#### Where should the reviewer start?
(Optional)

#### Any background context you want to provide?
(Optional)

#### Screenshots (if appropriate)
(Optional)

#### What GIF best describes this PR or how it makes you feel?
(Optional)

noisecapella commented:
This looks a little strange to me. Do we have a mockup for this yet? If not I'm fine with adjusting this in a separate PR 

Singleton:
![Screenshot from 2021-04-05 14-09-36](https://user-images.githubusercontent.com/863262/113609139-60f5c380-9619-11eb-91e6-5b30c48ecb9c.png)

Collection:
![Screenshot from 2021-04-05 14-09-04](https://user-images.githubusercontent.com/863262/113609155-67843b00-9619-11eb-9a67-149cc7e621cb.png)

noisecapella commented on static/js/components/SiteAddContent.tsx:
```diff
@@ -16,28 +16,65 @@ import {
+  const topLevelConfigItem: TopLevelConfigItem | null =
+    config.collections.find(
+      (config: TopLevelConfigItem) => config.name === type
+    ) || null
```
Can we use `??` instead of `||` here and below? It wouldn't make a difference in functionality here but it would be more explicit

gsidebo commented:
@noisecapella No mockup for this yet. If we had a similarly-styled "Add" button for the site metadata in your screenshot, it would need to be inline with the title. Our example config only has a single item in that "files" config section, so there's only one line, but remember that it can have multiple items defined, and they would all be rendered in that listing. If we had a fully-styled "Add" button on each of those lines, and those were mixed in with plain text "Edit" links, I'd argue that would be less consistent and less desirable. Consistency was my main goal with this UI

@Ferdi any thoughts on George's comment above?

noisecapella commented on static/js/components/SiteAddContent.tsx:
```diff
@@ -16,28 +16,65 @@ import {
+    return null
+  }
+  if (!subtype) {
+    if ("folder" in topLevelConfigItem) {
```
Can we use `isRepeatableCollectionItem` here?

noisecapella commented on static/js/query-configs/websites.ts:
```diff
@@ -235,57 +236,71 @@ export const createWebsiteCollaboratorMutation = (
-  offset: number
-): string => JSON.stringify([name, type, offset])
+  listingParams: ContentListingParams
+): string => JSON.stringify(Object.values(listingParams))
```
I think we need to specify the values to be stringified explicitly so that the key doesn't change with a different insertion order for `listingParams`. Otherwise you could have something like `[ name, type, offset ]` and `[type, name, offset]` which are two different keys for the same object.

noisecapella commented on static/js/util/factories/websites.ts:
```diff
@@ -38,25 +41,60 @@ export const makeWebsiteConfigField = (
+  name:     name || casual.word,
+  label:    casual.word,
+  category: casual.word,
+  ...(Math.random() >= 0.5 ?
```
Can we add a parameter which says which kind of config item this is, and use `Math.random` if the parameter is not set?  It would be useful to be help avoid flaky tests

Ferdi commented:
If the collection is of type "files" and there is only one file in that collection , it should show the form for that file 

![Side-nav – add-external – 1](https://user-images.githubusercontent.com/125561/113648973-915e5180-965b-11eb-8eb0-dfa5ec7aa6b0.png)

Every other file shows as an extra tab 

![Course-metadata-multi](https://user-images.githubusercontent.com/125561/113649036-bce13c00-965b-11eb-81bc-d15e83fea7b3.png)

gsidebo commented:
@Ferdi I'll add a new issue for those UI changes. This PR is pretty huge and we'd benefit from having it merged sooner rather than later

gsidebo commented:
@noisecapella Ready for review again

gsidebo commented:
@Ferdi Sorry, [this comment](https://github.com/mitodl/ocw-studio/pull/172#issuecomment-814134460) was way too brief and completely lacked context. I'm thinking that it would be best to merge this with the current UI and implement those suggestions in a separate PR mostly because this introduces a fundamental change in the config structure, and it's a big PR that touches a lot of the codebase so we risk merge conflicts the longer it stays open. 

Re: those suggestions, there are some implementation details we need to sort out:

> If the collection is of type "files" and there is only one file in that collection , it should show the form for that file

The wrinkle with this is that have a dedicated page for adding a new item, and a sidebar for editing an item. If we wanted to keep this same approach, we would need the "Site Metadata" link in our sidebar to either link to a new page or open the editing UI directly. To support this, we'd need to add an API call to fetch WebsiteContent records at the point when we're building the sidebar menu. We haven't had to do that up to this point. We just build the sidebar menu from the config.

> Every other file shows as an extra tab

Since a "files" config item can include an unlimited number of items, we may run into horizontal spacing issues if we use tabs.

noisecapella commented on websites/views.py:
```diff
@@ -256,10 +247,13 @@ class WebsiteContentViewSet(
         if filter_type:
             queryset = queryset.filter(type=filter_type)
+        if filter_subtype:
+            queryset = queryset.filter(subtype=filter_subtype)
```
Can you add a test for this? It looks like it's not covered by code coverage https://app.codecov.io/gh/mitodl/ocw-studio/compare/172/diff#D3L256

noisecapella commented on static/js/query-configs/websites.ts:
```diff
@@ -235,57 +236,82 @@ export const createWebsiteCollaboratorMutation = (
+    listingParams.offset
+  ]
+  if (listingParams.limit) {
+    key.push(listingParams.limit)
```
Can we add or modify a test so that it covers this line?

gsidebo replied:
This and the other lines you commented on in this file are tested by `static/js/components/SiteContentListing.test.tsx`

noisecapella replied:
I'm not sure, if I add a `throw new Error` at this line it doesn't change any test output. Same for the other lines

gsidebo replied:
That's odd... we're mocking the API request to return certain data given the URL with those extra parameters. I'll look into it

noisecapella commented on static/js/query-configs/websites.ts:
```diff
@@ -235,57 +236,82 @@ export const createWebsiteCollaboratorMutation = (
+  const { name, type, subtype, offset, limit } = listingParams
+  let url = siteApiContentListingUrl.param({ name }).query({ type, offset })
+  if (subtype) {
+    url = url.param({ subtype })
```
Can we add or modify a test so it covers this line?

noisecapella commented on static/js/query-configs/websites.ts:
```diff
@@ -235,57 +236,82 @@ export const createWebsiteCollaboratorMutation = (
+    url = url.param({ subtype })
+  }
+  if (limit) {
+    url = url.query({ limit })
```
Can we add or modify a test so it covers this line?

noisecapella commented on static/js/query-configs/websites.ts:
```diff
@@ -235,57 +236,82 @@ export const createWebsiteCollaboratorMutation = (
+        results = body.results
+      } else {
+        // If results are not paged, use dummy paging data.
+        pagingParams = { count: null, next: null, previous: null }
```
Can we add a test covering this case too?

noisecapella commented on static/js/components/SiteAddContent.tsx:
```diff
@@ -16,28 +19,66 @@ import {
+      return topLevelConfigItem as RepeatableConfigItem
+    }
+  } else if ("files" in topLevelConfigItem) {
+    return (
```
Can we add a test covering this condition?

noisecapella reviewed (changes requested):
Some missing tests. Functionality works great

Ferdi commented on websites/migrations/0017_content_subtype.py:
```diff
@@ -0,0 +1,18 @@
+    operations = [
+        migrations.AddField(
+            model_name="websitecontent",
+            name="subtype",
```
What is the need for this field ?

gsidebo replied:
Given a config that looks like this:

```yml
collections
  - name: navigation
    label: Navigation
    tag: Settings
    files:
      - file: config/menu.sidenav.yml
        label: Side Navigation
        name: sidenav
        fields: [...]
```

If you save a record for side navigation, that record would have "navigation" as the type and "sidenav" as the subtype

Ferdi replied:
Why is that field necessary?

gsidebo replied:
It's the field that relates the database record to the actual item in the config. In the authoring UI, we need a way to know if there is an existing record for one of those "file" items. In the example above, we would query the db for a record with type="navigation", subtype="sidenav". We could just set "sidenav" as the name, but then we wouldn't be able to query for all objects under "navigation", which is useful to determine which of the "file" items have an object created for them.

gsidebo replied:
We could potentially flip that and add a "parent_type" or something like that instead of "subtype". Then that example would have type="sidenav", parent_type="navigation"

noisecapella commented:
Just a thought, in a couple of those lines the condition tests for `limit`, but `limit` is always either `0` or `undefined`. Is the intention to test if `limit !== undefined || limit !== null` instead? I don't think we are sending `limit=0` to the server

gsidebo commented:
@noisecapella Great catch, you're absolutely right. I forgot about `if (0)`. Still works but we could run into issues with more than 1 page of results

gsidebo commented:
@noisecapella Just pushed a pretty big update to this PR and removed a bunch of the changes it introduced. I updated the PR description to reflect what this PR is doing now. Biggest change is that I removed the "subtype" field and there's no more setting the "limit" param in the API calls. This UI just lists the titles/labels for "file"-type items with no add/edit links, and you can't use the Add form for those item types either. That's just temporary and will be resolved in #192. 

Make sure you migrate down before pulling these changes (`manage.py migrate websites 0016`). This is ready for review again now. Sorry for the late pivot!

noisecapella commented:
FYI, more rebases incoming

gsidebo commented:
@noisecapella Should be ready now
